[25424] Implement frontend-side of relation columns

oliverguenther opened the pull request:
https://community.openproject.com/projects/openproject/work_packages/24454

**TODOS**

* [x] Don't handle clicks on additional rows (focus).
* [x] Stronger background color

oliverguenther committed:
Introduce derived state for when table query states are ready

oliverguenther committed:
Implement relation columns as secondary render pass

oliverguenther committed:
Extend backend relations API with WP type information

This saves the doubled frontend request just to find out which types
work packages of a specific relation have. Since the connection between work
package types and relation types are very close, we should always send
them.

oliverguenther committed:
Add relationType to relationOfType column

So we can avoid matching the id/href in the frontend.

oliverguenther committed:
Apply sorting check in huge_list

https://github.com/opf/openproject/pull/5635

oliverguenther committed:
Move QueryColumn interface into its own class

oliverguenther committed:
Render rows in timeline pass regardless of WP association

oliverguenther committed:
Fix timeline builder in inline-create row

oliverguenther committed:
Simplify initialization of query state

oliverguenther committed:
Add spec

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/485351a811475e2ec3e894e308966b75e8d5a328) (deployed commit: 7fa78d97ff92aafc0b19a8ad7042d558f318016d).

oliverguenther committed:
Only set scope to initialize table/timeline on initial loading

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/e73b94c5219ca1b1892faf6f54b88875b00e9598) (deployed commit: 831436191a4f14575956848df87acebafbf82b1e).

oliverguenther committed:
Add some happy debug outputs

oliverguenther committed:
Remove the loadingIndicator only after rendered was pushed

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/0244736ddd32675986ebf36cc29cd7519daa855d) (deployed commit: f82f1ab915a12059cea5cca538f49e2036e9f209).

oliverguenther committed:
Add better highlighting of expanded columns

oliverguenther committed:
Add better labels

oliverguenther committed:
Avoid setting the workPackageId to ignore clicks

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/d689e998cff8a2c577dc343dfabb665683c6cfcb) (deployed commit: 0e68018f0f3685065a3d61b2ef3bfdc5a364eda9).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/1c9d786412acaa0904e1c9514a39e0281151c934) (deployed commit: 569369092d9775d0d5789b2287b374d03fafc65c).

ulferts commented on app/models/queries/work_packages/columns/relation_of_type_column.rb:
```diff
@@ -45,6 +45,10 @@ def sym
 
+  def relation_type
+    sym
+  end
```
Why not simply use `sym` or at least alias it.

oliverguenther replied:
Done.

ulferts commented on frontend/app/components/api/api-v3/hal-resources/query-resource.service.ts:
```diff
@@ -77,15 +76,5 @@ function queryResource() {
-  _links?: {
-    self: { href:string, title:string };
-  }
-}
```
:+1: for turning it into a real HalResource. Never did get around to doing it.

ulferts commented on lib/api/v3/relations/relation_representer.rb:
```diff
@@ -52,6 +52,18 @@ class RelationRepresenter < ::API::Decorators::Single
+          {
+            href: api_v3_paths.type(represented.from.type_id)
+          }
+        end
```
I could be wrong but it looks like the code is no longer/not using those links. If so, I would be glad as they shouldn't be here IMO.

oliverguenther replied:
We're still using it here: https://github.com/opf/openproject/pull/5639/files#diff-d73e3377d1e7f10d8cdf091ce5cc523fR92

And since relations are tightly coupled to WP types in the frontend, we can use these links to make the relations tab much simpler without loading all work packages up front to do the sorting.

ulferts commented on frontend/app/components/api/api-work-packages/api-work-packages.service.ts:
```diff
@@ -60,6 +63,24 @@ export class ApiWorkPackagesService {
+   */
+  public loadWorkPackagesCollectionFor(ids:string[]) {
+    return this.halRequest.get(
+      '/api/v3/work_packages',
```
you could call `v3Path.wps()` here

oliverguenther replied:
Done.

ulferts commented on frontend/app/components/api/api-work-packages/api-work-packages.service.ts:
```diff
@@ -60,6 +63,24 @@ export class ApiWorkPackagesService {
+      '/api/v3/work_packages',
+      {
+        filters: JSON.stringify([{ id: {operator: '=', values: ids }}])
+      },
```
Not necessarily in this PR, but we should start using a service/helper for generating the filters.

oliverguenther replied:
Done.

ulferts replied:
👍  Thanks

ulferts commented on frontend/app/components/routing/wp-list/wp-list.controller.ts:
```diff
@@ -98,51 +97,55 @@ function WorkPackagesListController($scope:any,
+
+        // Update the page, if the change requires it
         if (triggerUpdate) {
           updateResultsVisibly(true);
```
The whole of the table reloads, including backend requests, when moving columns and I believe this change is the reason for it.

oliverguenther replied:
That's true. We don't currently figure out whether we _need_ to reload the table. We would need to compare the previous set of columns and decide whether this action requires a reload. Not sure if I can address this in the scope of this PR though.

ulferts reviewed (changes requested):
### Impression

I like almost all of the changes to the code with the exception of having the toType and fromType added which I consider bad design. I especially like the cleanup done to the states 👍 

Below are various bugs/suggestions:

#### Weird: Same content for different column type

It is probably desired behaviour but it strikes me as weird. Regardless of whether I expand a relationToType or a relationByType column, the output is always the same, prefixing the wp name with the relation type.  

relationByType:
![image](https://user-images.githubusercontent.com/617519/27134223-37491066-5115-11e7-89a5-8ec3af2a4fdc.png)

relationToType:
![image](https://user-images.githubusercontent.com/617519/27134443-de356848-5115-11e7-9bda-aad54d3f57a4.png)

If I expand a relationByType column, I already know the type relations listed there will have.

#### Bug: Moving column reloads table

Moving columns closes an expanded relation column. Not a serious issue IMO if at all but I seem to remember you talking about wanting to keep the expand/collapsed state (Now that I almost finished the review, I see that you are trying to keep the state). It is kept when (de)activating the hierarchy mode. Probably related to reloading the table on column move.

#### Bug: Row click area in hierarchy mode

Selecting rows has become very hard as clicking inside the area taken up by the inline edit block will not result in having the row selected. Message "Skipping click on inner cell". This is true only if the hierarchy mode is activated.

#### Bug:  Context menu on relation row

Right clicking on a relation row will result in a js error:
```
TypeError: Cannot read property 'object' of undefined
    at new wpContextMenuController (application.js:21730)
    at Object.s [as instantiate] (application.js:1223)
    at application.js:1224
    at u (application.js:1196)
    at application.js:1196
    at s (application.js:1225)
    at application.js:1225
    at f.$eval (application.js:1225)
    at f.$digest (application.js:1225)
    at application.js:1225
```

#### Bug: mouse icon on relation row

The mouse icon when hovering over the subject of a relation row is wrong (Text edit/selection icon) as it is not possible to edit the text.

#### Suggestion: different color for relations column

Probably highly subjective but the contrast between the two colors (relation column and selected column) is near to not existent, making the two colours hard to distinguish:
![image](https://user-images.githubusercontent.com/617519/27135269-63f5b698-5118-11e7-9eb2-3510771b3086.png)

oliverguenther committed:
Review remarks

oliverguenther committed:
Add very basic frontend filter builder

oliverguenther committed:
Print work package type for 'ofType' relation columns

oliverguenther committed:
Use separate dataLoaded state to mark table requirements loaded

This allows us to no longer clear the state of open/closed relation
columns when refreshing the table.

oliverguenther commented:
> **Bug: mouse icon on relation row**

Same as on additional hierarchy rows, and will be irrelevant after #25544

> **Bug: Row click area in hierarchy mode**

I don't think a change in this PR caused this. The fix made for 7.0.x is likely just not part of it yet.

> **Suggestion: different color for relations column**

The color is a simple variable so I suggest to move this away from this PR and let Niels/Henriette come up with something if they want to change it.

> **Bug: Moving column reloads table**

As mentioned in the comment above, we would need to check which actions require a table reload, that's not possible with the current change observers. Let's extract that into a separate task

oliverguenther commented:
@ulferts I have pushed fixes to all remarks we can address in this PR.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/82940d15ffa853689817337c7501672e8a9b1c67) (deployed commit: e7f58256848fd887a4895024bc742a08d33fbfbe).

oliverguenther committed:
Resolve transition after promise

oliverguenther committed:
Remove fromType, toType

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/173202969cd33dc39b4206fca04d221930b54909) (deployed commit: 507c2d26ea5f3a5d002bed7ead22dbd6a3d5c134).

oliverguenther committed:
Work on copy of renderedRows since we're modifying them

[ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/96ecf3891455a906b18b485604d46dd75052a205) (deployed commit: fea511397513fdb4c0a502265289705cfb91f541).
